Database synchronization with Slack

tomasfabry opened the pull request.

tomasfabry committed:
Synchronization frame

tomasfabry committed:
Bot channels loading

tomasfabry committed:
Channels new messages loading

tomasfabry committed:
Loading last channel message from DB

tomasfabry committed:
Adding new channel messages to DB

tomasfabry committed:
Synchronize all message in a channel after bot invitation

tomasfabry committed:
Performance optimalization

hleumas commented on bot/src/slack.js:
```diff
@@ -99,10 +103,123 @@ export const createOnMessageListener = (web) => {
+      const result = await web.users.conversations({
+        cursor,
+        limit: 999,
+        token: config.slack.botToken,
```
Why do we put a token explicitly here if we have already initialised slack with web token?

Also, why don't we import slack api in the slack.js file but do it in index.js?

Also, why do we use obsolete slack client module instead of recommended https://www.npmjs.com/package/@slack/web-api

?

Also, why don't we use  built-in pagination support in the client and instead emulate it ourselves?

https://www.npmjs.com/package/@slack/web-api#pagination

tomasfabry committed:
Fixed dependencies

Legacy dependency '@slack/client' replaced by '@slack/rtm-api' and '@slack/web-api'.

tomasfabry committed:
Clients initialization moved from 'index.js' to 'slack.js'

The reason is that we don't need sent client handle to every slack method as a parameter.

tomasfabry committed:
Tokens cleanup

There is no need to send the bot token to client methods because it was already set during client creation.

tomasfabry committed:
Used built-in pagination while retrieving Slack channel IDs

Manual pagination of Slack channel list replaced with built-in support for pagination.

tomasfabry committed:
Error handling moved to the top level

There is no need to handle slack errors on the low level so it was moved to the top level. Error during synchronization is logged only as an error message because the bot can run also without synchronization.

tomasfabry committed:
Fetching last report timestamp of all channel in one query.

Multiple queries for last report timestamps of all channels replaced by one query. In this query, all timestamps are saved into the map, where the channel ID is the key and the value is the timestamp of the latest report in this channel.
